ion/bitstream.go

```diff
@@ -692,32 +674,34 @@ func (b *bitstream) ReadTimestamp() (Timestamp, error) {
 }
 
 // ReadNsecs reads the fraction part of a timestamp and rounds to nanoseconds.
-// This function returns the nanoseconds as an int, overflow as a bool, and an error
+// This function returns the nanoseconds as an int, overflow as a bool, exponent as an uint8, and an error
 // if there was a problem executing this function.
-func (b *bitstream) readNsecs(length uint64) (int, bool, error) {
+func (b *bitstream) readNsecs(length uint64) (int, bool, uint8, error) {
 	d, err := b.readDecimal(length)
 	if err != nil {
-		return 0, false, err
+		return 0, false, 0, err
 	}
 
 	nsec, err := d.ShiftL(9).trunc()
 	if err != nil || nsec < 0 || nsec > 999999999 {
 		msg := fmt.Sprintf("invalid timestamp fraction: %v", d)
-		return 0, false, &SyntaxError{msg, b.pos}
+		return 0, false, 0, &SyntaxError{msg, b.pos}
 	}
 
 	nsec, err = d.ShiftL(9).round()
 	if err != nil {
 		msg := fmt.Sprintf("invalid timestamp fraction: %v", d)
-		return 0, false, &SyntaxError{msg, b.pos}
+		return 0, false, 0, &SyntaxError{msg, b.pos}
 	}
 
+	exponent := uint8(d.scale)
```

Comment:
Let's add a comment explaining that this cast should be safe because the max precision of 9 can fit inside uint8.

Reply:
That is not true because the [Ion timestamp spec](https://amzn.github.io/ion-docs/docs/spec.html#timestamp) mentions:

> Fractional seconds are allowed, with at least one digit of precision and an unlimited maximum.

Other Ion implementations can support arbitrary precision so Ion binary generated from Ion Java can have precision reduced.

Currently Ion Go support timestamps up to [9 precision units ](https://github.com/amzn/ion-go/releases/tag/v0.1.0-beta).

I created a [ticket](https://github.com/amzn/ion-go/issues/178) to update timestamps to support unlimited precision.